File: stipend_sign_flutter/lib/screens/profile_screen.dart
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:image_picker/image_picker.dart';
import '../models/user.dart';
import '../services/storage_service.dart';
import 'biometric_enrollment_screen.dart';

class ProfileScreen extends StatefulWidget {
  final User user;
  const ProfileScreen({super.key, required this.user});

  @override
  State<ProfileScreen> createState() => _ProfileScreenState();
}

class _ProfileScreenState extends State<ProfileScreen> {
  late final TextEditingController _firstNameCtrl;
  late final TextEditingController _lastNameCtrl;
  late final TextEditingController _studentIdCtrl;
  late final TextEditingController _departmentCtrl;
  late final TextEditingController _yearCtrl;
  late final TextEditingController _phoneCtrl;
  String? _cardImagePath;
  bool _loading = false;

  @override
  void initState() {
    super.initState();
    final u = widget.user;
    _firstNameCtrl = TextEditingController(text: u.firstName);
    _lastNameCtrl = TextEditingController(text: u.lastName);
    _studentIdCtrl = TextEditingController(text: u.studentId);
    _departmentCtrl = TextEditingController(text: u.department);
    _yearCtrl = TextEditingController(text: u.yearOfStudy);
    _phoneCtrl = TextEditingController(text: u.phone);
    _cardImagePath = u.studentCardImage;
  }

  @override
  void dispose() {
    for (final c in [
      _firstNameCtrl, _lastNameCtrl, _studentIdCtrl,
      _departmentCtrl, _yearCtrl, _phoneCtrl
    ]) {
      c.dispose();
    }
    super.dispose();
  }

  Future<void> _pickImage(ImageSource source) async {
    final picker = ImagePicker();
    final picked = await picker.pickImage(source: source, imageQuality: 80);
    if (picked != null) {
      setState(() => _cardImagePath = picked.path);
    }
  }

  void _showImageOptions() {
    showModalBottomSheet(
      context: context,
      builder: (_) => SafeArea(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            ListTile(
              leading: const Icon(Icons.camera_alt),
              title: const Text('Take Photo'),
              onTap: () {
                Navigator.pop(context);
                _pickImage(ImageSource.camera);
              },
            ),
            ListTile(
              leading: const Icon(Icons.photo_library),
              title: const Text('Choose from Gallery'),
              onTap: () {
                Navigator.pop(context);
                _pickImage(ImageSource.gallery);
              },
            ),
          ],
        ),
      ),
    );
  }

  Future<void> _handleSave() async {
    if (_firstNameCtrl.text.trim().isEmpty ||
        _lastNameCtrl.text.trim().isEmpty ||
        _studentIdCtrl.text.trim().isEmpty) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(
          content: Text('Please fill in all required fields'),
          backgroundColor: Colors.red,
        ),
      );
      return;
    }

    setState(() => _loading = true);

    try {
      final result = await StorageService.updateUser(widget.user.id, {
        'firstName': _firstNameCtrl.text.trim(),
        'lastName': _lastNameCtrl.text.trim(),
        'studentId': _studentIdCtrl.text.trim(),
        'department': _departmentCtrl.text.trim(),
        'yearOfStudy': _yearCtrl.text.trim(),
        'phone': _phoneCtrl.text.trim(),
        'studentCardImage': _cardImagePath,
      });

      if (result['success'] == true && mounted) {
        Navigator.of(context).pushReplacement(
          MaterialPageRoute(
            builder: (_) => BiometricEnrollmentScreen(user: result['user'] as User),
          ),
        );
      }
    } finally {
      if (mounted) setState(() => _loading = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Student Profile')),
      body: SingleChildScrollView(
        padding: const EdgeInsets.all(20),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            const Text(
              'Complete your profile and upload your student card',
              textAlign: TextAlign.center,
              style: TextStyle(color: Colors.grey, fontSize: 14),
            ),
            const SizedBox(height: 20),
            _sectionTitle('Personal Information'),
            _field(_firstNameCtrl, 'First Name *'),
            _field(_lastNameCtrl, 'Last Name *'),
            TextField(
              controller: _studentIdCtrl,
              enabled: widget.user.studentId.isEmpty,
              decoration: const InputDecoration(labelText: 'Student ID *'),
            ),
            const SizedBox(height: 12),
            _field(_departmentCtrl, 'Department'),
            _field(_yearCtrl, 'Year of Study (1-5)', type: TextInputType.number),
            _field(_phoneCtrl, 'Phone Number', type: TextInputType.phone),
            const SizedBox(height: 8),
            _sectionTitle('Student Card Photo'),
            const Text(
              'Upload a photo of your student ID card for verification',
              style: TextStyle(color: Colors.grey, fontSize: 13),
            ),
            const SizedBox(height: 10),
            GestureDetector(
              onTap: _showImageOptions,
              child: Container(
                height: 180,
                decoration: BoxDecoration(
                  border: Border.all(color: Colors.grey[300]!, width: 2),
                  borderRadius: BorderRadius.circular(10),
                ),
                child: _cardImagePath != null
                    ? ClipRRect(
                        borderRadius: BorderRadius.circular(10),
                        child: Image.file(File(_cardImagePath!), fit: BoxFit.cover),
                      )
                    : Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: const [
                          Icon(Icons.camera_alt, size: 50, color: Colors.grey),
                          SizedBox(height: 8),
                          Text('Tap to upload student card',
                              style: TextStyle(color: Colors.grey)),
                        ],
                      ),
              ),
            ),
            if (_cardImagePath != null) ...[
              const SizedBox(height: 8),
              TextButton(
                onPressed: () => setState(() => _cardImagePath = null),
                child: const Text('Remove Photo', style: TextStyle(color: Colors.red)),
              ),
            ],
            const SizedBox(height: 20),
            SizedBox(
              height: 50,
              child: ElevatedButton(
                onPressed: _loading ? null : _handleSave,
                style: ElevatedButton.styleFrom(
                  backgroundColor: const Color(0xFF4CAF50),
                  foregroundColor: Colors.white,
                  shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10)),
                ),
                child: _loading
                    ? const SizedBox(
                        height: 20,
                        width: 20,
                        child: CircularProgressIndicator(color: Colors.white, strokeWidth: 2),
                      )
                    : const Text('Save & Continue',
                        style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
              ),
            ),
            const SizedBox(height: 30),
          ],
        ),
      ),
    );
  }

  Widget _sectionTitle(String title) => Padding(
        padding: const EdgeInsets.only(bottom: 10, top: 8),
        child: Text(title,
            style: const TextStyle(
                fontSize: 18, fontWeight: FontWeight.bold, color: Color(0xFF4CAF50))),
      );

  Widget _field(TextEditingController ctrl, String label,
          {TextInputType type = TextInputType.text}) =>
      Padding(
        padding: const EdgeInsets.only(bottom: 12),
        child: TextField(controller: ctrl, keyboardType: type, decoration: InputDecoration(labelText: label)),
      );
}

File: sign in stipend app/screens/ProfileScreen.js
// filepath: screens/ProfileScreen.js
import React, { useState, useEffect } from 'react';
import { View, Text, TextInput, TouchableOpacity, StyleSheet, Alert, Image, ScrollView } from 'react-native';
import * as ImagePicker from 'expo-image-picker';
import { updateUser } from '../utils/Storage';
import BackButton from '../components/BackButton';

export default function ProfileScreen({ navigation, route }) {
  const [user, setUser] = useState(route.params?.user || null);
  const [formData, setFormData] = useState({
    firstName: user?.firstName || '',
    lastName: user?.lastName || '',
    studentId: user?.studentId || '',
    department: user?.department || '',
    yearOfStudy: user?.yearOfStudy || '',
    phone: user?.phone || '',
  });
  const [studentCardImage, setStudentCardImage] = useState(user?.studentCardImage || null);
  const [loading, setLoading] = useState(false);

  const updateField = (field, value) => {
    setFormData(prev => ({ ...prev, [field]: value }));
  };

  const pickImage = async () => {
    const permissionResult = await ImagePicker.requestMediaLibraryPermissionsAsync();
    
    if (!permissionResult.granted) {
      Alert.alert('Permission Required', 'Please allow access to your photo library.');
      return;
    }

    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [4, 3],
      quality: 0.8,
    });

    if (!result.canceled) {
      setStudentCardImage(result.assets[0].uri);
    }
  };

  const takePhoto = async () => {
    const permissionResult = await ImagePicker.requestCameraPermissionsAsync();
    
    if (!permissionResult.granted) {
      Alert.alert('Permission Required', 'Please allow access to your camera.');
      return;
    }

    const result = await ImagePicker.launchCameraAsync({
      allowsEditing: true,
      aspect: [4, 3],
      quality: 0.8,
    });

    if (!result.canceled) {
      setStudentCardImage(result.assets[0].uri);
    }
  };

  const showImageOptions = () => {
    Alert.alert(
      'Upload Student Card',
      'Choose an option',
      [
        { text: 'Take Photo', onPress: takePhoto },
        { text: 'Choose from Gallery', onPress: pickImage },
        { text: 'Cancel', style: 'cancel' },
      ]
    );
  };

  const handleSave = async () => {
    if (!formData.firstName || !formData.lastName || !formData.studentId) {
      Alert.alert('Error', 'Please fill in all required fields');
      return;
    }

    setLoading(true);
    
    try {
      const result = await updateUser(user.id, {
        ...formData,
        studentCardImage: studentCardImage,
      });
      
      if (result.success) {
        Alert.alert('Success', 'Profile saved!', [
          { text: 'OK', onPress: () => navigation.replace('BiometricEnrollment', { user: result.user }) }
        ]);
      } else {
        Alert.alert('Error', result.message);
      }
    } catch (error) {
      Alert.alert('Error', 'Something went wrong');
    }
    
    setLoading(false);
  };

  return (
    <ScrollView style={styles.container}>
      <BackButton navigation={navigation} />
      <Text style={styles.header}>Student Profile</Text>
      <Text style={styles.subtitle}>Please complete your profile and upload your student card</Text>

      <View style={styles.section}>
        <Text style={styles.sectionTitle}>Personal Information</Text>
        
        <TextInput
          style={styles.input}
          placeholder="First Name *"
          value={formData.firstName}
          onChangeText={(v) => updateField('firstName', v)}
        />
        
        <TextInput
          style={styles.input}
          placeholder="Last Name *"
          value={formData.lastName}
          onChangeText={(v) => updateField('lastName', v)}
        />
        
        <TextInput
          style={styles.input}
          placeholder="Student ID *"
          value={formData.studentId}
          onChangeText={(v) => updateField('studentId', v)}
          editable={!user?.studentId}
        />
        
        <TextInput
          style={styles.input}
          placeholder="Department"
          value={formData.department}
          onChangeText={(v) => updateField('department', v)}
        />
        
        <TextInput
          style={styles.input}
          placeholder="Year of Study (1-5)"
          value={formData.yearOfStudy}
          onChangeText={(v) => updateField('yearOfStudy', v)}
          keyboardType="numeric"
        />
        
        <TextInput
          style={styles.input}
          placeholder="Phone Number"
          value={formData.phone}
          onChangeText={(v) => updateField('phone', v)}
          keyboardType="phone-pad"
        />
      </View>

      <View style={styles.section}>
        <Text style={styles.sectionTitle}>Student Card Photo</Text>
        <Text style={styles.helpText}>Upload a photo of your student ID card for verification</Text>
        
        <TouchableOpacity style={styles.imageButton} onPress={showImageOptions}>
          {studentCardImage ? (
            <Image source={{ uri: studentCardImage }} style={styles.cardImage} />
          ) : (
            <View style={styles.imagePlaceholder}>
              <Text style={styles.imagePlaceholderText}>📷</Text>
              <Text style={styles.imagePlaceholderLabel}>Tap to upload student card</Text>
            </View>
          )}
        </TouchableOpacity>
        
        {studentCardImage && (
          <TouchableOpacity style={styles.removeButton} onPress={() => setStudentCardImage(null)}>
            <Text style={styles.removeButtonText}>Remove Photo</Text>
          </TouchableOpacity>
        )}
      </View>

      <TouchableOpacity 
        style={styles.button} 
        onPress={handleSave}
        disabled={loading}
      >
        <Text style={styles.buttonText}>
          {loading ? 'Saving...' : 'Save & Continue'}
        </Text>
      </TouchableOpacity>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    padding: 20,
  },
  header: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#333',
    textAlign: 'center',
  },
  subtitle: {
    fontSize: 14,
    color: '#666',
    textAlign: 'center',
    marginBottom: 20,
  },
  section: {
    marginBottom: 25,
  },
  sectionTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    color: '#4CAF50',
    marginBottom: 10,
  },
  helpText: {
    fontSize: 14,
    color: '#666',
    marginBottom: 10,
  },
  input: {
    borderWidth: 1,
    borderColor: '#ddd',
    borderRadius: 10,
    padding: 15,
    marginBottom: 12,
    fontSize: 16,
  },
  imageButton: {
    borderWidth: 2,
    borderColor: '#ddd',
    borderRadius: 10,
    borderStyle: 'dashed',
    overflow: 'hidden',
  },
  imagePlaceholder: {
    height: 200,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#f9f9f9',
  },
  imagePlaceholderText: {
    fontSize: 50,
  },
  imagePlaceholderLabel: {
    marginTop: 10,
    color: '#666',
    fontSize: 14,
  },
  cardImage: {
    width: '100%',
    height: 200,
    resizeMode: 'cover',
  },
  removeButton: {
    marginTop: 10,
    alignItems: 'center',
  },
  removeButtonText: {
    color: '#f44336',
    fontSize: 14,
  },
  button: {
    backgroundColor: '#4CAF50',
    padding: 15,
    borderRadius: 10,
    alignItems: 'center',
    marginTop: 10,
    marginBottom: 40,
  },
  buttonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
});